pubtools/pulplib/_impl/model/repository/repo_lock.py

```diff
@@ -0,0 +1,163 @@
+import json
+import random
+import string
+from time import sleep
+import os
+import logging
+
+from datetime import datetime, timedelta
+from more_executors.futures import f_map
+
+LOCK_CLAIM_STR = "pulplib-lock-claim-"
+
+LOCK_VALID_FROM_OFFSET = os.getenv("PULPLIB_LOCK_VALID_FROM_OFFSET", 10)
+LOCK_EXPIRATION_OFFSET = os.getenv("PULPLIB_LOCK_EXPIRATION_OFFSET", 1800)
+
+AWAIT_ACTIVATION_SLEEP = os.getenv("PULPLIB_AWAIT_ACTIVATION_SLEEP", 10)
+
+REPO_ENDPOINT_BASE = "pulp/api/v2/repositories/%s/"
+
+LOG = logging.getLogger("pubtools.pulplib")
+
+
+class LockClaim(object):
+    def __init__(self, expiration_offset=LOCK_EXPIRATION_OFFSET, **kwargs):
+        self.id = (
+            kwargs["id"]
+            if "id" in kwargs
+            else "".join(random.choices(string.ascii_letters + string.digits, k=8))
+        )
+        self.created = (
+            datetime.fromtimestamp(kwargs["created"])
+            if "created" in kwargs
+            else datetime.now()
+        )
+        self.valid_from = (
+            datetime.fromtimestamp(kwargs["valid_from"])
+            if "valid_from" in kwargs
+            else self.created + timedelta(seconds=LOCK_VALID_FROM_OFFSET)
+        )
+        self.expires = (
+            datetime.fromtimestamp(kwargs["expires"])
+            if "expires" in kwargs
+            else self.created + timedelta(seconds=expiration_offset)
+        )
+
+    @property
+    def as_json(self):
+        data = {
+            "id": self.id,
+            "created": datetime.timestamp(self.created),
+            "valid_from": datetime.timestamp(self.valid_from),
+            "expires": datetime.timestamp(self.expires),
+        }
+        return json.dumps(data)
+
+    def __lt__(self, other):
+        return self.created < other.created
+
+    def __eq__(self, other):
+        return self.id == other.id
+
+    @staticmethod
+    def from_json_data(json_data):
+        return LockClaim(**json_data)
+
+    @property
+    def is_valid(self):
+        return self.expires > datetime.now()
+
+
+class RepoLock(object):
+    def __init__(
+        self,
+        repo_id,
+        request_func,
+        expiration_offset=LOCK_EXPIRATION_OFFSET,
+        activation_await_sleep=AWAIT_ACTIVATION_SLEEP,
+    ):
+        self._repo_id = repo_id
+        self._request_func = request_func
+        self._lock_claim = None
+        self._expiration_offset = expiration_offset
+        self._activation_await_sleep = activation_await_sleep
+
+    def __enter__(self):
+        LOG.info("Attempting to lock %s." % self._repo_id)
+        self.remove_expired_locks()
+        self._lock_claim = LockClaim(expiration_offset=self._expiration_offset)
+        self.submit_lock_claim()
+        self.await_lock_activation()
+
+    def __exit__(self, exc_type, exc_value, tb):
+        self.delete_lock_claim()
```

Comment:
While I wouldn't call it absolutely necessary, there is a bit of a problem right now in this scenario:

- caller requests a lock of duration X
- the time between enter/exit ends up being *longer than* X

i.e. the caller requested a too low duration.  This means the caller didn't actually hold the lock for as long as they needed it.

While this is the caller's fault for requesting a too low duration, and we can't prevent it, maybe we should at least detect it?  e.g. at exit time, should we do one of the following?

1. check if current time is less than the expiry time; or...
2. fetch the current locks and verify that our lock still exists (before deleting it)

...and then log a warning?  Otherwise, in cases where a client requests too small duration, there'll be no trace of this happening - just silent data corruption, the same kind of corruption we're trying to fix.

Option (1) seems pretty easy to implement so maybe it's worthwhile.  e.g. it could generate a message along the lines of:

`ERROR: client requested Pulp repo lock for $X seconds but held lock for $Y seconds!`

It's definitely a bug so ERROR seems right, though not sure it's severe enough to raise.